File: challenges/Algorithmic/Highest prime factor/HighPF.hs
{-# LANGUAGE RecordWildCards #-}

{-%
Highest Prime Factor (Haskell)
==============================

Command line utility mirroring the Python implementation in the same directory.
Computes the largest prime factor for one or more integers, supporting positional
arguments, batch file inputs, stdin streaming, JSON output, and optional timing.
-}

module Main (main, highestPrimeFactor) where

import Control.Exception (IOException, try)
import Data.List (intercalate)
import Data.Time.Clock (diffUTCTime, getCurrentTime)
import System.Environment (getArgs)
import System.Exit (ExitCode (..), exitWith)
import System.IO (hPutStrLn, stderr)
import Text.Printf (printf)
import Text.Read (readMaybe)

-- | Compute the largest prime factor of @n@ using a 6k ± 1 stride trial division.
highestPrimeFactor :: Integer -> Integer
highestPrimeFactor n
  | n < 2 = error "Input must be an integer greater than 1."
  | otherwise = loop after3 best3 5
  where
    (after2, best2) = stripPrime n 2 1
    (after3, best3) = stripPrime after2 3 best2

    loop :: Integer -> Integer -> Integer -> Integer
    loop value best f
      | value == 1 = best
      | f * f > value = max value best
      | otherwise =
          let (value1, best1) = stripPrime value f best
              (value2, best2) = stripPrime value1 (f + 2) best1
           in loop value2 best2 (f + 6)

    stripPrime :: Integer -> Integer -> Integer -> (Integer, Integer)
    stripPrime value factor best =
      let (reduced, found) = stripFactor value factor
          newBest = if found then max best factor else best
       in (reduced, newBest)

    stripFactor :: Integer -> Integer -> (Integer, Bool)
    stripFactor value factor = go value False
      where
        go current seen
          | current `mod` factor == 0 = go (current `div` factor) True
          | otherwise = (current, seen)

-- | Result for one integer, optionally carrying elapsed seconds.
data FactorResult = FactorResult
  { frValue :: Integer
  , frHighest :: Integer
  , frElapsed :: Maybe Double
  }

-- | Options parsed from the CLI.
data Options = Options
  { optNumbers :: [Integer]
  , optUseStdin :: Bool
  , optBatchFiles :: [FilePath]
  , optJSON :: Bool
  , optTiming :: Bool
  }
  deriving (Show)

usageMessage :: String
usageMessage = unlines
  [ "Usage: HighPF [--stdin] [--batch FILE ...] [--json] [--timing] [NUMBERS ...]"
  , "Compute highest prime factors for integers >= 2." ]

emptyOptions :: Options
emptyOptions = Options [] False [] False False

parseArgs :: [String] -> Either String Options
parseArgs = go emptyOptions
  where
    go opts [] = Right opts
    go opts ("--stdin" : rest) = go opts {optUseStdin = True} rest
    go opts ("--json" : rest) = go opts {optJSON = True} rest
    go opts ("--timing" : rest) = go opts {optTiming = True} rest
    go opts ("--batch" : path : rest) = go opts {optBatchFiles = path : optBatchFiles opts} rest
    go _ ("--batch" : []) = Left "--batch expects a file path argument."
    go opts (flag@('-' : '-') : _) = Left $ "Unrecognised option: " ++ flag
    go opts (value : rest) =
      case readMaybe value of
        Just n -> go opts {optNumbers = optNumbers opts ++ [n]} rest
        Nothing -> Left $ "Invalid integer argument: " ++ value

loadBatchFile :: FilePath -> IO (Either String [Integer])
loadBatchFile path = do
  result <- try (readFile path) :: IO (Either IOException String)
  case result of
    Left err -> pure . Left $ "Unable to read batch file " ++ path ++ ": " ++ show err
    Right contents -> pure $ parseNumberStream ("file: " ++ path) contents

parseNumberStream :: String -> String -> Either String [Integer]
parseNumberStream label raw = traverse parseToken tokens
  where
    normalised = map replaceComma raw
    tokens = words normalised
    parseToken token =
      case readMaybe token of
        Just n -> Right n
        Nothing -> Left $ "Invalid integer in " ++ label ++ ": " ++ token
    replaceComma ',' = ' '
    replaceComma c = c

processNumbers :: Bool -> [Integer] -> IO [FactorResult]
processNumbers timing numbers = traverse compute numbers
  where
    compute value = do
      start <- if timing then Just <$> getCurrentTime else pure Nothing
      let highest = highestPrimeFactor value
      end <- if timing then Just <$> getCurrentTime else pure Nothing
      let elapsed = case (start, end) of
            (Just s, Just e) -> Just (realToFrac (diffUTCTime e s))
            _ -> Nothing
      pure FactorResult {frValue = value, frHighest = highest, frElapsed = elapsed}

renderResults :: Bool -> [FactorResult] -> String
renderResults True [] = "[]"
renderResults True results =
  let renderItem FactorResult {..} =
        "  {" ++ intercalate ", " (baseFields ++ timingField) ++ " }"
        where
          baseFields =
            [ "\"value\": " ++ show frValue
            , "\"highest_prime_factor\": " ++ show frHighest
            ]
          timingField =
            case frElapsed of
              Just t -> ["\"elapsed_seconds\": " ++ showFFloat t]
              Nothing -> []
      body = intercalate ",\n" (map renderItem results)
   in "[\n" ++ body ++ "\n]"
renderResults False results = unlines (map renderLine results)
  where
    renderLine FactorResult {..} =
      case frElapsed of
        Just t -> printf "%d: %d (elapsed %.3f ms)" frValue frHighest (t * 1000)
        Nothing -> printf "%d: %d" frValue frHighest

showFFloat :: Double -> String
showFFloat = printf "%.9f"

main :: IO ()
main = do
  args <- getArgs
  case parseArgs args of
    Left err -> do
      hPutStrLn stderr usageMessage
      hPutStrLn stderr $ "[error] " ++ err
      exitWith (ExitFailure 2)
    Right opts -> runWith opts

runWith :: Options -> IO ()
runWith opts@Options {..} = do
  stdinNumbers <-
    if optUseStdin
      then do
        contents <- getContents
        case parseNumberStream "stdin" contents of
          Left err -> quit 2 err
          Right nums -> pure nums
      else pure []
  batchNumbers <- traverse loadBatchFile (reverse optBatchFiles)
  case sequence batchNumbers of
    Left err -> quit 2 err
    Right batches -> do
      let numbers = optNumbers ++ stdinNumbers ++ concat batches
      if null numbers
        then do
          hPutStrLn stderr usageMessage
          hPutStrLn stderr "No numbers provided (use positional args, --batch, or --stdin)."
          exitWith (ExitFailure 1)
        else
          case filter (< 2) numbers of
            (bad : _) -> quit 3 $ "All numbers must be >= 2 (got " ++ show bad ++ ")."
            [] -> do
              results <- processNumbers optTiming numbers
              putStrLn $ renderResults optJSON results
              exitWith ExitSuccess
  where
    quit code message = do
      hPutStrLn stderr $ "[error] " ++ message
      exitWith (ExitFailure code)

File: challenges/Algorithmic/Highest prime factor/HighPF.py
"""Highest Prime Factor (modernized)
===================================

Compute the largest prime factor of one or more positive integers efficiently.

Enhancements vs original version:
 - Optimized 6k±1 trial division after stripping factors of 2 and 3.
 - Handles very large integers (limited only by memory/time) using pure Python integers.
 - Supports multiple inputs via CLI, JSON output, and optional timing.
 - Clear error handling and exit codes (non-zero on invalid input).
 - Well documented and type annotated for new contributors.

Time complexity: O( sqrt(n)/log n ) practical due to factor elimination (worst-case for large prime).

Example usage:
  python HighPF.py 13195 600851475143
  python HighPF.py --json 42 97 10
  echo 600851475143 | python HighPF.py --stdin --timing
"""

from __future__ import annotations

import argparse
import json
import sys
import time
from dataclasses import dataclass
from typing import Iterable, List, Sequence


def highest_prime_factor(n: int) -> int:
    """Returns the largest prime factor of a given integer.

    This function uses an optimized trial division method. It first handles
    the factors of 2 and 3, and then checks for factors of the form 6k ± 1
    up to the square root of the remaining number.

    Args:
        n: An integer greater than 1.

    Returns:
        The largest prime factor of `n`.

    Raises:
        ValueError: If `n` is less than 2.
    """
    if n < 2:
        raise ValueError("Input must be an integer greater than 1.")

    # Track current best factor
    max_factor = 1  # will be replaced; n>=2 guarantees answer >=2

    # Remove all factors of 2
    while n % 2 == 0:
        n //= 2
        max_factor = 2
    # Remove all factors of 3
    while n % 3 == 0:
        n //= 3
        max_factor = 3  # overwrites if 3 divides

    # Now test potential factors of form 6k ± 1
    f = 5
    # We can stop when f^2 > n; when loop exits if n>1 it is prime.
    while f * f <= n:
        if n % f == 0:
            max_factor = f
            while n % f == 0:
                n //= f
        # Check companion f+2 (since between 6k-1 and 6k+1 there is one even and one divisible by 3)
        f2 = f + 2
        if n % f2 == 0:
            max_factor = f2
            while n % f2 == 0:
                n //= f2
        f += 6

    if n > 1:  # Remaining n is prime
        max_factor = n

    return int(max_factor)


@dataclass
class FactorResult:
    value: int
    highest_prime_factor: int
    elapsed_seconds: float | None = None


def process_numbers(numbers: Sequence[int], timing: bool = False) -> List[FactorResult]:
    """Processes a sequence of numbers to find their highest prime factors.

    Args:
        numbers: A sequence of integers to process.
        timing: If `True`, measures the time taken to process each number.

    Returns:
        A list of `FactorResult` objects.
    """
    results: List[FactorResult] = []
    for num in numbers:
        start = time.perf_counter() if timing else None
        hpf = highest_prime_factor(num)
        elapsed = (
            (time.perf_counter() - start) if timing and start is not None else None
        )
        results.append(FactorResult(num, hpf, elapsed))
    return results


def parse_stdin() -> List[int]:
    """Parses a list of integers from standard input.

    The input can be a single line of space- or comma-separated integers.

    Returns:
        A list of integers parsed from stdin.

    Raises:
        ValueError: If any of the tokens in the input cannot be converted to an integer.
    """
    data = sys.stdin.read().strip()
    if not data:
        return []
    out: List[int] = []
    for token in data.replace(",", " ").split():
        try:
            out.append(int(token))
        except ValueError:
            raise ValueError(f"Invalid integer in stdin input: {token}") from None
    return out


def build_arg_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(
        description="Compute highest prime factor for one or more integers."
    )
    p.add_argument("numbers", nargs="*", type=int, help="Integers >= 2")
    p.add_argument(
        "--stdin",
        action="store_true",
        help="Also read whitespace/comma separated integers from stdin",
    )
    p.add_argument("--json", action="store_true", help="Output results as JSON array")
    p.add_argument(
        "--timing", action="store_true", help="Include elapsed seconds per number"
    )
    return p


def render_results(results: Sequence[FactorResult], as_json: bool) -> str:
    if as_json:
        return json.dumps(
            [
                {
                    "value": r.value,
                    "highest_prime_factor": r.highest_prime_factor,
                    **(
                        {"elapsed_seconds": r.elapsed_seconds}
                        if r.elapsed_seconds is not None
                        else {}
                    ),
                }
                for r in results
            ],
            indent=2,
        )
    # Plain text output
    lines = []
    for r in results:
        base = f"{r.value}: {r.highest_prime_factor}"
        if r.elapsed_seconds is not None:
            base += f" (elapsed {r.elapsed_seconds * 1000:.3f} ms)"
        lines.append(base)
    return "\n".join(lines)


def main(argv: Iterable[str] | None = None) -> int:
    parser = build_arg_parser()
    args = parser.parse_args(list(argv) if argv is not None else None)

    numbers: List[int] = list(args.numbers)
    if args.stdin:
        try:
            numbers.extend(parse_stdin())
        except ValueError as e:
            print(f"[error] {e}", file=sys.stderr)
            return 2

    if not numbers:
        parser.print_usage(sys.stderr)
        print("No numbers provided (use positional args or --stdin).", file=sys.stderr)
        return 1

    # Validate inputs
    for n in numbers:
        if n < 2:
            print(f"[error] All numbers must be >= 2 (got {n}).", file=sys.stderr)
            return 3

    try:
        results = process_numbers(numbers, timing=args.timing)
    except ValueError as e:  # from highest_prime_factor, defensive
        print(f"[error] {e}", file=sys.stderr)
        return 4

    print(render_results(results, as_json=args.json))
    return 0


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())
